src/math/quat.zig

```diff
@@ -15,9 +16,256 @@ pub fn Quat(comptime Scalar: type) type {
+                by = -by;
+                bz = -bz;
+                bw = -bw;
+            }
+
+            var scale0: T = 0.0;
+            var scale1: T = 0.0;
+
+            if (1.0 - cosOmega > math.eps(T)) {
+                const omega = std.math.acos(cosOmega);
+                const sinOmega = std.math.sin(omega);
+                scale0 = std.math.sin((1.0 - t) * omega) / sinOmega;
+                scale1 = std.math.sin(t * omega) / sinOmega;
+            } else {
+                scale0 = 1.0 - t;
+                scale1 = t;
+            }
+
+            return init(scale0 * ax + scale1 * bx, scale0 * ay + scale1 * by, scale0 * az + scale1 * bz, scale0 * aw + scale1 * bw);
+        }
+
+        /// Calculates the conjugate of the given quaternion.
+        pub inline fn conjugate(q: *const Quat(T)) Quat(T) {
+            return init(-q.v.x(), -q.v.y(), -q.v.z(), q.v.w());
+        }
+
+        /// Creates a quaternion from the given transformation matrix.
+        pub inline fn fromMat(comptime matT: type, m: *const matT) Quat(T) {
+            var dst = Quat(T).identity();
+            const trace = m.v[0].v[0] + m.v[1].v[1] + m.v[2].v[2];
+
+            if (trace > 0) {
+                const root = std.math.sqrt(trace + 1.0);
+                dst.v.v[3] = 0.5 * root;
+                const rootInv = 0.5 / root;
+
+                dst.v.v[0] = (m.v[1].v[2] - m.v[2].v[1]) * rootInv;
+                dst.v.v[1] = (m.v[2].v[0] - m.v[0].v[2]) * rootInv;
+                dst.v.v[2] = (m.v[0].v[1] - m.v[1].v[0]) * rootInv;
+            } else {
+                var i: usize = 0;
+
+                if (m.v[1].v[1] > m.v[0].v[0]) {
+                    i = 1;
+                }
+
+                if (m.v[2].v[2] > m.v[i].v[i]) {
+                    i = 2;
+                }
+
+                const j = (i + 1) % 3;
+                const k = (i + 2) % 3;
+
+                const root = std.math.sqrt(m.v[i].v[i] - m.v[j].v[j] - m.v[k].v[k] + 1.0);
+                dst.v.v[i] = 0.5 * root;
+
+                const rootInv = 0.5 / root;
+
+                dst.v.v[3] = (m.v[j].v[k] - m.v[k].v[j]) * rootInv;
+                dst.v.v[j] = (m.v[j].v[i] - m.v[i].v[j]) * rootInv;
+                dst.v.v[k] = (m.v[k].v[i] - m.v[i].v[k]) * rootInv;
+            }
+
+            return dst;
+        }
+
+        /// Creates a quaternion from the given Euler angles.
+        pub inline fn fromEuler(x: T, y: T, z: T) Quat(T) {
+            const xHalf = x * 0.5;
+            const yHalf = y * 0.5;
+            const zHalf = z * 0.5;
+
+            const sx = std.math.sin(xHalf);
+            const cx = std.math.cos(xHalf);
+            const sy = std.math.sin(yHalf);
+            const cy = std.math.cos(yHalf);
+            const sz = std.math.sin(zHalf);
+            const cz = std.math.cos(zHalf);
+
+            return init(sx * cy * cz + cx * sy * sz, cx * sy * cz - sx * cy * sz, cx * cy * sz + sx * sy * cz, cx * cy * cz - sx * sy * sz);
```

Comment:
Could we split the x, y, z, w components out into constants above this for slightly improved readability?

Reply:
yeah, I assume you want the same in similar places?